com.unity.ml-agents/Runtime/Actuators/ActuatorManager.cs

```diff
@@ -0,0 +1,317 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace Unity.MLAgents.Actuators
+{
+    /// <summary>
+    /// A class that manages the delegation of events and data structures for IActuators
+    /// </summary>
+    public class ActuatorManager : IList<IActuator>
+    {
+        IList<IActuator> m_Actuators;
+        int m_SumOfDiscreteBranchSizes;
+        int m_NumDiscreteBranches;
+        int m_ContinuousSize;
+
+        /// <summary>
+        /// Create an ActuatorList with a preset capacity.
+        /// </summary>
+        /// <param name="capacity">The capacity of the list to create.</param>
+        public ActuatorManager(int capacity = 0)
+        {
+            m_Actuators = new List<IActuator>(capacity);
+        }
+
+        /// <summary>
+        /// Returns the previously stored actions for the actuators in this list.
+        /// </summary>
+        public float[] StoredContinuousActions { get; private set; }
+
+        /// <summary>
+        /// Returns the previously stored actions for the actuators in this list.
+        /// </summary>
+        public int[] StoredDiscreteActions { get; private set; }
+
+        public int TotalNumberOfActions { get; private set; }
+
+        BufferedDiscreteActionMask m_DiscreteActionMask;
+        public IDiscreteActionMask DiscreteActionMask => m_DiscreteActionMask;
+
+        /// <summary>
+        /// Allocations the action buffer and ensures that its size matches the number of actions
+        /// these collection of IActuators can execute.
+        /// </summary>
+        public void EnsureActionBufferSize()
```

Comment:
So the user needs to call this? I'm worried about making it clear if they try to use it without calling, and also what happens if they call it and also modify the actuators afterwards (so the Stored*Actions will be too small).

What about
* add a flag indicated EnsureActionBufferSize() has been called, and return if the flag is set (so EnsureActionBufferSize is idempotent)
* throw if Add() is called after the flag is set
* Clear() also clears the flag (maybe, not sure)
* ExecuteActions() calls EnsureActionBufferSize()
?

Reply:
yeah, i also didn't like that it was a requirement to call from the user of the class.  I like the idea of it being called from execute actions.